state_synchronizer/src/peer_manager.rs

```diff
@@ -102,20 +108,39 @@ impl PeerManager {
                     peer_info.score = new_score.max(MIN_SCORE);
                 }
             }
+            if (old_score - peer_info.score).abs() > std::f64::EPSILON {
+                self.compute_weighted_index();
+            }
         }
     }
 
-    pub fn pick_peer(&mut self) -> Option<(PeerId, StateSynchronizerSender)> {
+    fn compute_weighted_index(&mut self) {
         let active_peers = self.get_active_upstream_peers();
-        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
-
         if !active_peers.is_empty() {
             let weights: Vec<_> = active_peers
                 .iter()
                 .map(|(_, peer_info)| peer_info.score)
                 .collect();
+            match WeightedIndex::new(&weights) {
+                Ok(weighted_index) => {
+                    self.weighted_index = Some(weighted_index);
+                }
+                Err(e) => {
+                    error!(
+                        "[state sync] (pick_peer) failed to compute weighted index, {:?}",
+                        e
+                    );
+                }
+            }
+        }
+    }
+
+    pub fn pick_peer(&self) -> Option<(PeerId, StateSynchronizerSender)> {
+        let active_peers = self.get_active_upstream_peers();
+        debug!("[state sync] (pick_peer) state: {:?}", self.peers);
 
-            if let Ok(weighted_index) = WeightedIndex::new(&weights) {
+        if !active_peers.is_empty() {
```

Comment:
do we need this check?

Reply:
I think so because we fetch using index later? `active_peers[weighted_index.sample(&mut rng)]`

Reply:
we actually don't need this specific check. But it bothers me a bit that if active_peers and weighted_index will go out of sync, we'd panic on line 145

Reply:
Hm maybe I should check that `weighted_index.len() == active_peers.len()`? And recompute weighted index if they are out of sync for some reason. That check wouldn't guarantee that each index refers to the correct peer, but it at least wouldn't panic

Reply:
just use `.get` api instead of `[..]` call

Reply:
I agree with the concern - if it is expected that `weighted_index.len() == active_peers.len()`, could we wrap the 2 within a common struct? We might then be able to use MIRAI macros to assert the invariant.

Reply:
Updated the function to use .get and removed active.is_empty check. Think that should be sufficient for now, let me know your thoughts